### ISSUE TITLE
[question] Is there any way to get the task queue size (backlog)?

### ISSUE THREAD

**Issue by tarampampam** [temporalio/sdk-go#1043]
Hi there and thank you so much for the great service!

My question may be a bit stupid, but anyway - is there any legal way to get the task queue size? I mean: "How many tasks in the queue `%name%` are scheduled for processing, but have not been processed yet?".

This is necessary to implement the autoscaling mechanism (more tasks in queue for processing - more workers, and vice versa) on my side. And yes, I know about the "[custom workers metrics](https://github.com/temporalio/samples-go/tree/main/metrics)", but I am looking for a better (read - simpler) way.

**Comment by cretz** [comment]
You can use `Client.DescribeTaskQueue` which is a low-level call that returns https://pkg.go.dev/go.temporal.io/api/workflowservice/v1#DescribeTaskQueueResponse that contains a status with a backlog count hint, but it may be a fuzzy number. You can do the same via tctl: https://docs.temporal.io/tctl-v1/taskqueue/#describe.

> This is necessary to implement the autoscaling mechanism

I am not sure that is the best way to scale workers. A better way is to check whether the workers have slots of work available to still perform. If they don't, you need more workers usually. See https://docs.temporal.io/application-development/worker-performance.

You can also feel free to come join us on the forums or public Slack to discuss worker tuning.

**Comment by dnr** [comment]
Take a look at these docs also: https://docs.temporal.io/server/production-deployment#faqs
In general we recommend scaling workers based on `schedule_to_start_latency`.

**Comment by tarampampam** [comment]
Thanks a lot for your answers, guys! I'll back later with additional questions or close this issue as resolved, ok?

**Comment by tarampampam** [comment]
```go
// I have tried with `video_transcoding` too
resp, rErr := temporal.Client().DescribeTaskQueue(ctx, "video-transcoding", enums.TASK_QUEUE_TYPE_ACTIVITY)
if rErr != nil {
	return rErr
}

fmt.Println(resp.TaskQueueStatus)
```

Is always `nil`, but... Why? A tasks queue exists (I have workers on it when I execute `DescribeTaskQueue`).

And about the `schedule_to_start_latency` metric:

```txt
task_schedule_to_start_latency_sum{namespace="default",operation="RecordActivityTaskStarted",service_name="history",task_type="Activity",taskqueue="default"} 4.012997128000003
task_schedule_to_start_latency_count{namespace="default",operation="RecordActivityTaskStarted",service_name="history",task_type="Activity",taskqueue="default"} 314
task_schedule_to_start_latency_sum{namespace="default",operation="RecordActivityTaskStarted",service_name="history",task_type="Activity",taskqueue="video_transcoding"} 421589.4196587619
task_schedule_to_start_latency_count{namespace="default",operation="RecordActivityTaskStarted",service_name="history",task_type="Activity",taskqueue="video_transcoding"} 75
task_schedule_to_start_latency_sum{namespace="default",operation="RecordWorkflowTaskStarted",service_name="history",task_type="Workflow",taskqueue="__sticky__"} 4.952772867000001
task_schedule_to_start_latency_count{namespace="default",operation="RecordWorkflowTaskStarted",service_name="history",task_type="Workflow",taskqueue="__sticky__"} 362
temporal_activity_schedule_to_start_latency_sum{activity_type="temporal_sys_tq_scanner_scvg_activity",client_name="temporal_go",namespace="temporal_system",service_name="worker",task_queue="temporal_sys_tq_scanner_taskqueue_0",worker_type="none",workflow_type="temporal_sys_tq_scanner_workflow"} 0.008792396
temporal_activity_schedule_to_start_latency_count{activity_type="temporal_sys_tq_scanner_scvg_activity",client_name="temporal_go",namespace="temporal_system",service_name="worker",task_queue="temporal_sys_tq_scanner_taskqueue_0",worker_type="none",workflow_type="temporal_sys_tq_scanner_workflow"} 1
```

It looks like static, because when I have ~10 scheduled activities on the `video-transcoding` without workers on it, and when workers are started to process those jobs - nothing was changed :(

**Comment by cretz** [comment]
> Is always nil, but... Why?

I think you'll have to call `Client.WorkflowService().DescribeTaskQueue(` with a full https://pkg.go.dev/go.temporal.io/api/workflowservice/v1#DescribeTaskQueueRequest that has `IncludeTaskQueueStatus` as `true` to get status. Usually doing a low-level task queue describe is not used for scaling, there are other methods. See the worker tuning guide I mentioned.

I'll defer to @dnr on `schedule_to_start_latency` and the server FAQ.

**Comment by tarampampam** [comment]
@cretz Thanks for your answer! But it still doesn't work for me :( At this moment I have many scheduled activities in the task queue named `video-transcoding`:

![image](https://user-images.githubusercontent.com/7326800/219673769-5dd99cdd-d737-4702-ab6b-d4c658c71e68.png)

![image](https://user-images.githubusercontent.com/7326800/219675015-4b4e2015-f4d5-4566-b8ea-1fe5b3681903.png)

But without running workers for these activities - they should be in the backlog, right? To make a decision to up or downscale the workers (attached to this task queue) I read the server metrics, and as said in the documentation:

> Poll Success Rate = (poll_success + poll_success_sync) / (poll_success + poll_success_sync + poll_timeouts)
>
> Poll Success Rate should be >90% in most cases of systems with a steady load. For high volume and low latency, try to target >95%.

I got: `{PollSuccess:72 PollSuccessSync:50 PollTimeouts:16}; Poll Success Rate: 88.40579710144928`. After a set of experiments, I noticed this value does not depend on the actual workers' count (all tests were made on my laptop).

An attempt to get the backlog size is also failed:

```go
resp, tErr := temporal.Client().WorkflowService().DescribeTaskQueue(ctx, &workflowservice.DescribeTaskQueueRequest{
  Namespace: opt.temporal.namespace,
  TaskQueue: &v14.TaskQueue{
    Name: "video-transcoding",
  },
  TaskQueueType: enums.TASK_QUEUE_TYPE_ACTIVITY,
  IncludeTaskQueueStatus: true,
})

if tErr != nil {
  return tErr
}

log.Info("Task queue", zap.Any("resp", resp))
```

In the logs:

```
Task queue      {"resp": "&DescribeTaskQueueResponse{Pollers:[]*PollerInfo{},TaskQueueStatus:&v14.TaskQueueStatus{BacklogCountHint:0,ReadLevel:700000,AckLevel:700000,RatePerSecond:1000,TaskIdBlock:&TaskIdBlock{StartId:800001,EndId:900000,},},}"}
```

As you can see `BacklogCountHint` is `0` (always). What am I doing wrong? 😞

**Comment by tarampampam** [comment]
Guys?

**Comment by cretz** [comment]
> But without running workers for these activities - they should be in the backlog, right?

Yes, but the hint may not reflect that value. I am getting details on that value now, but it is a mostly internal value that you should not use for scaling workers. Please do not use this for scaling and ignore this low-level API task queue status hint for your use case.

> To make a decision to up or downscale the workers (attached to this task queue) I read the server metrics, and as said in the documentation

I think that part of the documentation you were looking at is only for downscaling (i.e. having too many workers). To know whether to scale up, you should use the available slots metric for knowing when to scale up (or change options).

Simply put - you can use server-side poll metrics to see if there are more workers polling than there are tasks to be given. This will let you know you have too many (which is mostly harmless). You can use SDK-side slot metrics to know whether the slots are being all used up to know whether you need to scale up.

If this is unclear in https://docs.temporal.io/application-development/worker-performance, can we improve it in some way?

Also, for more back-and-forth communication like questions, feel free to ask in our forums or Slack.
